### ISSUE TITLE
continued pre-training / more sequences in the OAS dataset

### ISSUE THREAD

**Issue by avilella** [chandar-lab/AMPLIFY#23]
Hi,

I am interested in taking one of the existing models, and continuing the training from one of the checkpoint files by adding more data of the kind found in the OAS dataset. I intend to add about another 1 million pairs of chains like the ones in the oas_train dataset.

Is there a recipe to do this I could apply? I have access to 2xA100 / 4xA100 or 2xH100 nodes for this (re the bf16 precision), but I can also do the testing on a VRAM 48Gb A6000 GPU.

Thanks in advance

**Comment by qfournier** [comment]
Hi @avilella,

Thank you for reaching out. Yes, this is definitely possible! There are two approaches you can consider.

1. **Using Our Codebase (Not Recommended):** You could create a job that loads the checkpoints and continues training in a manner similar to how we extended training with longer sequences. However, the code expects `accelerate` checkpoints with optimizer states, which we did not release as they require the exact same number of GPUs (either 8 or 32 in our case). As a result, you would need to modify our code to load a weight-only checkpoint and train “from scratch”.

2. **Using Hugging Face (Recommended):** This method is a lot more straightforward. You can load the model as you would typically do for inference and use the [Hugging Face Trainer](https://huggingface.co/docs/transformers/en/main_classes/trainer) to continue training. The model will effectively resume training.

While we haven’t explored continual pre-training yet, here are some best practices to consider:

- Learning Rate: Warm up the learning rate again, as the optimizer states have been reset. Keep the learning rate small (around 10^-4) to prevent moving away from the current minima and preserve the knowledge learned. You should track the performance on our UniProt validation set to make sure you are not degrading the model's knowledge too much.
- Batch Size: Use a batch size at least as large as what was used during the initial pre-training (2M tokens), as larger batch sizes have been shown to help, especially towards the end of training. Since you have access to 2 x A100 / H100, you can use gradient accumulation to reach the desired batch size.
- Data Formatting: Ensure the chains are separated by the correct “|” character, and verify that the padding mask is additive rather than multiplicative.

For context, AMPLIFY 120M achieved 93.8% accuracy and a perplexity of 1.27 on the OAS validation set at the end of 1 million training steps.

I hope this helps.  

Best,
 Quentin
